#### Parte 1: Carga y Exploración de Datos

**Carga del Conjunto de Datos**

Instalar dependencias

In [16]:
%pip install pandas
%pip install db-dtypes

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


Importar pandas e importar el dataset

In [17]:
import pandas as pd

data_path="../data/datos.csv"

df =pd.read_csv(data_path)
df

productAsin country        date  isVerified  ratingScore  \
0     B09G9BL5CP   India  11-08-2024        True            4   
1     B09G9BL5CP   India  16-08-2024        True            5   
2     B09G9BL5CP   India  14-05-2024        True            4   
3     B09G9BL5CP   India  24-06-2024        True            5   
4     B09G9BL5CP   India  18-05-2024        True            5   
...          ...     ...         ...         ...          ...   
3057  B09G9D8KRQ   India  11-10-2023        True            1   
3058  B09G9D8KRQ   India  14-10-2022        True            2   
3059  B09G9D8KRQ   India  24-02-2022        True            3   
3060  B09G9D8KRQ   India  16-10-2023        True            1   
3061  B09G9D8KRQ   India  11-11-2023        True            1   

                              reviewTitle  \
0                              No charger   
1                         iPhone 13 256GB   
2                 Flip camera option nill   
3                                 Product   
4                            Good product   
...                                   ...   
3057  Very bad experience with i phone 13   
3058    not happy with this apple product   
3059                           Good phone   
3060                    Battery discharge   
3061    Batter power needs to be improved   

                                      reviewDescription  \
0     Every thing is good about iPhones, there's not...   
1     It look so fabulous, I am android user switche...   
2     I tried to flip camera while recording but no ...   
3                                          100% genuine   
4            Happy to get the iPhone 13 in Amazon offer   
...                                                 ...   
3057  Useless phon never buy this heat n useless cam...   
3058  iam not happy with this product why because ch...   
3059                                         Good phone   
3060  While charging mobile it's getting so hot even...   
3061  Battery power is be very bad need to chat on d...   

                                              reviewUrl  \
0     https://www.amazon.in/gp/customer-reviews/R345...   
1     https://www.amazon.in/gp/customer-reviews/R2HJ...   
2     https://www.amazon.in/gp/customer-reviews/R3Y7...   
3     https://www.amazon.in/gp/customer-reviews/R1P9...   
4     https://www.amazon.in/gp/customer-reviews/R1XI...   
...                                                 ...   
3057  https://www.amazon.in/gp/customer-reviews/R10O...   
3058  https://www.amazon.in/gp/customer-reviews/R2FW...   
3059  https://www.amazon.in/gp/customer-reviews/R2C7...   
3060  https://www.amazon.in/gp/customer-reviews/R3K0...   
3061  https://www.amazon.in/gp/customer-reviews/R2QO...   

                                 reviewedIn  \
0       Reviewed in India on 11 August 2024   
1       Reviewed in India on 16 August 2024   
2          Reviewed in India on 14 May 2024   
3         Reviewed in India on 24 June 2024   
4          Reviewed in India on 18 May 2024   
...                                     ...   
3057   Reviewed in India on 11 October 2023   
3058   Reviewed in India on 14 October 2022   
3059  Reviewed in India on 24 February 2022   
3060   Reviewed in India on 16 October 2023   
3061  Reviewed in India on 11 November 2023   

                                variant variantAsin  
0          Colour: MidnightSize: 256 GB  B09G9BQS98  
1          Colour: MidnightSize: 256 GB  B09G9BQS98  
2          Colour: MidnightSize: 256 GB  B09G9BQS98  
3          Colour: MidnightSize: 256 GB  B09G9BQS98  
4          Colour: MidnightSize: 256 GB  B09G9BQS98  
...                                 ...         ...  
3057  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3058  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3059  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3060  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  
3061  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N  

[3062 rows x 11 columns]

Se identifican las columnas reviewDescription que contiene el texto de la reseña y ratingScore que indica la valoracion, segun la valoracion podemos identificar si la reseña es positiva o negativa.

**Descripción Inicial de los Datos**

Se cuenta la cantidad de reseñas presentes

In [18]:
num_filas = df.shape[0]
print(f"El DataFrame tiene {num_filas} filas.")

El DataFrame tiene 3062 filas.


El dataframe tiene  3062 reseñas

Ahora veremos si hay mas reseñas positivas o negativas. En nuestro caso, basado en la columna **ratingScore**, positivo sera mayor que 3 y menor o igual a 3 sera negativo. Para diferenciarlo mejor, añadiremos una columna booleana que contendra un *Positive* o *Negative*.

In [19]:
df['ratingPositive'] = df['ratingScore'].apply(lambda x: 'Positive' if x > 3 else 'Negative')

df.to_csv("../data/datosConColumnaPositivos.csv",index=False)
df

productAsin country        date  isVerified  ratingScore  \
0     B09G9BL5CP   India  11-08-2024        True            4   
1     B09G9BL5CP   India  16-08-2024        True            5   
2     B09G9BL5CP   India  14-05-2024        True            4   
3     B09G9BL5CP   India  24-06-2024        True            5   
4     B09G9BL5CP   India  18-05-2024        True            5   
...          ...     ...         ...         ...          ...   
3057  B09G9D8KRQ   India  11-10-2023        True            1   
3058  B09G9D8KRQ   India  14-10-2022        True            2   
3059  B09G9D8KRQ   India  24-02-2022        True            3   
3060  B09G9D8KRQ   India  16-10-2023        True            1   
3061  B09G9D8KRQ   India  11-11-2023        True            1   

                              reviewTitle  \
0                              No charger   
1                         iPhone 13 256GB   
2                 Flip camera option nill   
3                                 Product   
4                            Good product   
...                                   ...   
3057  Very bad experience with i phone 13   
3058    not happy with this apple product   
3059                           Good phone   
3060                    Battery discharge   
3061    Batter power needs to be improved   

                                      reviewDescription  \
0     Every thing is good about iPhones, there's not...   
1     It look so fabulous, I am android user switche...   
2     I tried to flip camera while recording but no ...   
3                                          100% genuine   
4            Happy to get the iPhone 13 in Amazon offer   
...                                                 ...   
3057  Useless phon never buy this heat n useless cam...   
3058  iam not happy with this product why because ch...   
3059                                         Good phone   
3060  While charging mobile it's getting so hot even...   
3061  Battery power is be very bad need to chat on d...   

                                              reviewUrl  \
0     https://www.amazon.in/gp/customer-reviews/R345...   
1     https://www.amazon.in/gp/customer-reviews/R2HJ...   
2     https://www.amazon.in/gp/customer-reviews/R3Y7...   
3     https://www.amazon.in/gp/customer-reviews/R1P9...   
4     https://www.amazon.in/gp/customer-reviews/R1XI...   
...                                                 ...   
3057  https://www.amazon.in/gp/customer-reviews/R10O...   
3058  https://www.amazon.in/gp/customer-reviews/R2FW...   
3059  https://www.amazon.in/gp/customer-reviews/R2C7...   
3060  https://www.amazon.in/gp/customer-reviews/R3K0...   
3061  https://www.amazon.in/gp/customer-reviews/R2QO...   

                                 reviewedIn  \
0       Reviewed in India on 11 August 2024   
1       Reviewed in India on 16 August 2024   
2          Reviewed in India on 14 May 2024   
3         Reviewed in India on 24 June 2024   
4          Reviewed in India on 18 May 2024   
...                                     ...   
3057   Reviewed in India on 11 October 2023   
3058   Reviewed in India on 14 October 2022   
3059  Reviewed in India on 24 February 2022   
3060   Reviewed in India on 16 October 2023   
3061  Reviewed in India on 11 November 2023   

                                variant variantAsin ratingPositive  
0          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
1          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
2          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
3          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
4          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive  
...                                 ...         ...            ...  
3057  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative  
3058  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative  
3059  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative  
3060  Colo

Contamos la cantidad de *Positive* y *Negative*

In [20]:
conteo_positivos_negativos = df['ratingPositive'].value_counts()

print(conteo_positivos_negativos)

ratingPositive
Positive    2065
Negative     997
Name: count, dtype: int64


Hay 2065 reseñas positivas y 997 negativas.

#### Parte 2: Preprocesamiento de Texto

**Preprocesamiento del Texto**

Instalar nltk y spacy

In [28]:
%pip install nltk
%pip install spacy
!python -m spacy download en_core_web_sm

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     ---- ----------------------------------- 1.6/12.8 MB 8.4 MB/s eta 0:00:02
     --------------- ------------------------ 5.0/12.8 MB 13.1 MB/s eta 0:00:01
     ------------------------------------ -- 12.1/12.8 MB 20.4 MB/s eta 0:00:01
     --------------------------------------- 12.8/12.8 MB 20.6 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


Tokenización  
Ademas se quitan espacios vacios y caracteres molestos como "," y "."

In [29]:
import spacy

nlp = spacy.load("en_core_web_sm")

df['reviewDescription'] = df['reviewDescription'].fillna('')

df['preprocessedReview'] = ''

df['preprocessedReview'] = df['reviewDescription'].apply(lambda doc: [token.text for token in nlp(doc) if not token.is_punct and not token.is_space and token.text not in ['"', "'"]])

df

productAsin country        date  isVerified  ratingScore  \
0     B09G9BL5CP   India  11-08-2024        True            4   
1     B09G9BL5CP   India  16-08-2024        True            5   
2     B09G9BL5CP   India  14-05-2024        True            4   
3     B09G9BL5CP   India  24-06-2024        True            5   
4     B09G9BL5CP   India  18-05-2024        True            5   
...          ...     ...         ...         ...          ...   
3057  B09G9D8KRQ   India  11-10-2023        True            1   
3058  B09G9D8KRQ   India  14-10-2022        True            2   
3059  B09G9D8KRQ   India  24-02-2022        True            3   
3060  B09G9D8KRQ   India  16-10-2023        True            1   
3061  B09G9D8KRQ   India  11-11-2023        True            1   

                              reviewTitle  \
0                              No charger   
1                         iPhone 13 256GB   
2                 Flip camera option nill   
3                                 Product   
4                            Good product   
...                                   ...   
3057  Very bad experience with i phone 13   
3058    not happy with this apple product   
3059                           Good phone   
3060                    Battery discharge   
3061    Batter power needs to be improved   

                                      reviewDescription  \
0     Every thing is good about iPhones, there's not...   
1     It look so fabulous, I am android user switche...   
2     I tried to flip camera while recording but no ...   
3                                          100% genuine   
4            Happy to get the iPhone 13 in Amazon offer   
...                                                 ...   
3057  Useless phon never buy this heat n useless cam...   
3058  iam not happy with this product why because ch...   
3059                                         Good phone   
3060  While charging mobile it's getting so hot even...   
3061  Battery power is be very bad need to chat on d...   

                                              reviewUrl  \
0     https://www.amazon.in/gp/customer-reviews/R345...   
1     https://www.amazon.in/gp/customer-reviews/R2HJ...   
2     https://www.amazon.in/gp/customer-reviews/R3Y7...   
3     https://www.amazon.in/gp/customer-reviews/R1P9...   
4     https://www.amazon.in/gp/customer-reviews/R1XI...   
...                                                 ...   
3057  https://www.amazon.in/gp/customer-reviews/R10O...   
3058  https://www.amazon.in/gp/customer-reviews/R2FW...   
3059  https://www.amazon.in/gp/customer-reviews/R2C7...   
3060  https://www.amazon.in/gp/customer-reviews/R3K0...   
3061  https://www.amazon.in/gp/customer-reviews/R2QO...   

                                 reviewedIn  \
0       Reviewed in India on 11 August 2024   
1       Reviewed in India on 16 August 2024   
2          Reviewed in India on 14 May 2024   
3         Reviewed in India on 24 June 2024   
4          Reviewed in India on 18 May 2024   
...                                     ...   
3057   Reviewed in India on 11 October 2023   
3058   Reviewed in India on 14 October 2022   
3059  Reviewed in India on 24 February 2022   
3060   Reviewed in India on 16 October 2023   
3061  Reviewed in India on 11 November 2023   

                                variant variantAsin ratingPositive  \
0          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
1          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
2          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
3          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
4          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
...                                 ...         ...            ...   
3057  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative   
3058  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative   
3059  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative   


In [31]:
df.to_csv("../data/a.csv",index=False)

Quitar Stop Words

In [32]:
df['preprocessedReview'] = df['preprocessedReview'].apply(lambda tokens: [word for word in tokens if not nlp.vocab[word].is_stop])

df

productAsin country        date  isVerified  ratingScore  \
0     B09G9BL5CP   India  11-08-2024        True            4   
1     B09G9BL5CP   India  16-08-2024        True            5   
2     B09G9BL5CP   India  14-05-2024        True            4   
3     B09G9BL5CP   India  24-06-2024        True            5   
4     B09G9BL5CP   India  18-05-2024        True            5   
...          ...     ...         ...         ...          ...   
3057  B09G9D8KRQ   India  11-10-2023        True            1   
3058  B09G9D8KRQ   India  14-10-2022        True            2   
3059  B09G9D8KRQ   India  24-02-2022        True            3   
3060  B09G9D8KRQ   India  16-10-2023        True            1   
3061  B09G9D8KRQ   India  11-11-2023        True            1   

                              reviewTitle  \
0                              No charger   
1                         iPhone 13 256GB   
2                 Flip camera option nill   
3                                 Product   
4                            Good product   
...                                   ...   
3057  Very bad experience with i phone 13   
3058    not happy with this apple product   
3059                           Good phone   
3060                    Battery discharge   
3061    Batter power needs to be improved   

                                      reviewDescription  \
0     Every thing is good about iPhones, there's not...   
1     It look so fabulous, I am android user switche...   
2     I tried to flip camera while recording but no ...   
3                                          100% genuine   
4            Happy to get the iPhone 13 in Amazon offer   
...                                                 ...   
3057  Useless phon never buy this heat n useless cam...   
3058  iam not happy with this product why because ch...   
3059                                         Good phone   
3060  While charging mobile it's getting so hot even...   
3061  Battery power is be very bad need to chat on d...   

                                              reviewUrl  \
0     https://www.amazon.in/gp/customer-reviews/R345...   
1     https://www.amazon.in/gp/customer-reviews/R2HJ...   
2     https://www.amazon.in/gp/customer-reviews/R3Y7...   
3     https://www.amazon.in/gp/customer-reviews/R1P9...   
4     https://www.amazon.in/gp/customer-reviews/R1XI...   
...                                                 ...   
3057  https://www.amazon.in/gp/customer-reviews/R10O...   
3058  https://www.amazon.in/gp/customer-reviews/R2FW...   
3059  https://www.amazon.in/gp/customer-reviews/R2C7...   
3060  https://www.amazon.in/gp/customer-reviews/R3K0...   
3061  https://www.amazon.in/gp/customer-reviews/R2QO...   

                                 reviewedIn  \
0       Reviewed in India on 11 August 2024   
1       Reviewed in India on 16 August 2024   
2          Reviewed in India on 14 May 2024   
3         Reviewed in India on 24 June 2024   
4          Reviewed in India on 18 May 2024   
...                                     ...   
3057   Reviewed in India on 11 October 2023   
3058   Reviewed in India on 14 October 2022   
3059  Reviewed in India on 24 February 2022   
3060   Reviewed in India on 16 October 2023   
3061  Reviewed in India on 11 November 2023   

                                variant variantAsin ratingPositive  \
0          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
1          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
2          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
3          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
4          Colour: MidnightSize: 256 GB  B09G9BQS98       Positive   
...                                 ...         ...            ...   
3057  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative   
3058  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative   
3059  Colour: (PRODUCT) REDSize: 128 GB  B09G99CW2N       Negative   


In [34]:
df.to_csv("../data/a.csv",index=False)